crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,707 @@
+use std::marker::PhantomData;
+
+use bevy_utils::{get_short_name, HashSet};
+
+use crate::{
+    component::{ComponentId, Components},
+    prelude::Bundle,
+    world::World,
+};
+
+/// A rule about which [`Component`](crate::component::Component)s can coexist on entities.
+///
+/// These rules must be true at all times for all entities in the [`World`].
+/// The generic [`Bundle`] type `B1` is always used in the `premise`,
+/// while `B2` is used in the `consequence`.
+/// If only a single generic is provided, these types are the same.
+///
+/// When added to the [`World`], archetype invariants behave like [`assert!`].
+/// Archetype invariants are checked each time [`Archetypes`](crate::archetype::Archetypes) is modified;
+/// this can occur on component addition, component removal, and entity spawning.
+///
+/// Archetypes are only modified when a novel archetype (set of components) is seen for the first time;
+/// swapping between existing archetypes will not trigger these checks.
+///
+/// Note that archetype invariants are not symmetric by default.
+/// For example, `ArchetypeInvariant::<B1, B2>::requires_one()` means that `B1` requires `B2`,
+/// but not that `B2` requires `B1`.
+/// In this case, an entity with just `B2` is completely valid, but an entity with just `B1` is not.
+/// If symmetry is desired, repeat the invariant with the order of the types switched.
```

Comment:
Yep, this is the correct impl.

Reply:
That looks like the correct impl, but I don't think we want to add this method for two reasons:
1. When not using inverse, users only have to swap the order of the generic parameters, which is just as easy as calling a new method in my opinion.
2. For many common invariants, reversing the premise and consequence isn't actually what's desired for something that's conceptually "reversed". For example, `requires` uses "`AnyOf(B1)` implies `AllOf(B2)`". Swapping the premise and consequence would produce "`AllOf(B2)` implies `AnyOf(B1)`", when the likely desired reverse would actually be "`AnyOf(B2)` implies `AllOf(B1)`".

Reply:
Hmm, I also didn't think about the case where `B1` and `B2` are different: the return type isn't `Self` since the generic type arguments are reversed. This is the `cargo check`ed (still not tested) version:

```rust
/// Returns an `ArchetypeInvariant` from this where premise and consequence are swapped.
#[inline]
pub fn inverse(&self) -> ArchetypeInvariant<B2, B1>
where
    B1: Clone,
    B2: Clone,
{
    ArchetypeInvariant {
        premise: self.consequence.clone(),
        consequence: self.premise.clone(),
    }
}
```

I don't know though if we should add it, since to make it useful, the user must write the original one as a variable to then use it twice in the app building phase.

Reply:
Agreed. This might be a nice piece of sugar, but I think the juice just isn't quite worth the squeeze in this case.